### ISSUE TITLE
Opening inventory is extremely slow

### ISSUE THREAD

**Issue by Shad000w** [Beamdog/nwn-issues#368]
**Specifics**

 - build .28 but this is not new, was happening from start
 - OS: windows10

**If needed, describe the bug**

Opening inventory of any object, be it creature or placeable or store is extremely slow, It is even worse when want builder to open or change properties of some item that is in palette from the inventory window. Not only the item properties windows takes long to open, but after changing the item it produces about 30second lag before inventory of the object loads again.

It often leads to one of the windows showing up behind others which then means the toolset has to be shutdown via task manager to continue working.

**Comment by virusman** [comment]
@Shad000w Is this reproducible with an empty module without haks?

**Comment by Shad000w** [comment]
It doesn't. I am not sure what exactly causes this, might be number of item bluprints, because I tested it in SP module using CEP2 and the issue wasn't there. So basically it is happening only with my PW module of 110MB size and 8302 item blueprints.

**Comment by virusman** [comment]
If you can point to a module that I can download where this also happens, that'd be very helpful.

**Comment by Shad000w** [comment]
Try World of Rhun again

https://neverwintervault.org/sites/all/modules/pubdlcnt/pubdlcnt.php?fid=34525

But the freeze when opening inventory/creating a blueprint copy while an inventory is opened in this module is very short it is there. This module has just 2k item blueprints compared to my PW with 8k. Perhaps generating 8k random item blueprints o reproduce the issue would help, but I am not sure how to do that.

**Comment by pF-arQon** [comment]
Can very much confirm this bug. We "only" have ~2500 custom items in our PW, but opening any creature's inventory takes 20-30s, and as Shad000w says, gods forbid you actually edit one of those items, because the *module's* (not "creature's") entire item list is clearly being reparsed *every* time the inventory list is redrawn.

Editing the items directly from the palette is quick: it's only within the Creature Properties dialog that the code is falling apart. (Apparently either running an O^^2 loop of some kind, or re-serialising the UTI from disk for every item every time: it'll all be in cache, but repeatedly re-parsing the object's nested variants every time could easily account for how cripplingly slow it is).

**Comment by tinygiant98** [comment]
@pF-arQon Do you have a way to provide a repro module where we can examine this behavior more closely? 20s seems excessive. Also, confirm it's still occurring on .35-36.

**Comment by pF-arQon** [comment]
No, but anything with enough items will show it. Note that Shad000w's 30s is more than 20s, so...  :)

It was still present in 35, and it actually goes back to at least 1.69 - I tested in that sometime last year.

Shad000w's sort-of mistaken about it not being reproducible in "simple" modules without CC: the behavior is clearly still *there*, it's just that in a toy module where the whole thing fits in L1 and O is tiny, the same actions "only" take a second or two.

Where this gets interesting is that if you run the toolset in WINE on a Linux box, inventories only take a couple of seconds to open even with our thousands of additional items. Where it gets *really* interesting is that if you then run the TS in a W10 VM, it *still* only takes a couple of seconds.

So, that pretty solidly points at a couple of likely causes, with a garbage allocator and/or broken VMM use the easy front-runners. Re-parsing the UTIs every time still seems a plausible candidate, since that would also involve millions of de/allocs while destroying the entire on-chip cache as well; but given the Forms code involved it could just be something as stupid as resizing an array on every insertion.
